search-commons/src/main/java/no/unit/nva/search2/common/QueryBuilder.java

```diff
@@ -0,0 +1,182 @@
+        }
+        if (!invalidKeys.isEmpty()) {
+            throw new BadRequestException(validQueryParameterNamesMessage(invalidKeys, validKeys()));
+        }
+        notValidated = false;
+        return this;
+    }
+
+    /**
+     * Adds query and path parameters from requestInfo.
+     */
+    public final QueryBuilder<T, U> fromRequestInfo(RequestInfo requestInfo) {
+        query.gatewayUri = requestInfo.getRequestUri();
+        return fromQueryParameters(requestInfo.getQueryParameters());
+    }
+
+
+    /**
+     * Adds parameters from query.
+     */
+    public QueryBuilder<T, U> fromQueryParameters(Map<String, String> parameters) {
+        parameters.forEach(this::setValue);
+        return this;
+    }
+
+
+    /**
+     * Defines which parameters are required.
+     * @param requiredParameters comma seperated QueryParameterKeys
+     */
+    @SafeVarargs
+    public final QueryBuilder<T, U> withRequiredParameters(T... requiredParameters) {
+        var tmpSet = Set.of(requiredParameters);
+        query.otherRequiredKeys.addAll(tmpSet);
+        return this;
+    }
+
+    /**
+     * Sample code for assignDefaultValues.
+     * <p>Usage:</p>
+     * <samp>requiredMissing().forEach(key -> { <br>
+     *     switch (key) {<br>
+     *         case LANGUAGE:<br>
+     *             query.setValue(key, DEFAULT_LANGUAGE_CODE);<br>
+     *             break;<br>
+     *         default:<br>
+     *             break;<br>
+     *     }});<br>
+     * </samp>
+     */
+    protected abstract void assignDefaultValues();
+
+    /**
+     * Sample code for setValue.
+     * <p>Usage:</p>
+     * <samp>var qpKey = keyFromString(key,value);<br>
+     * if(qpKey.equals(INVALID)) {<br>
+     *     invalidKeys.add(key);<br>
+     * } else {<br>
+     *     query.setValue(qpKey, value);<br>
+     * }<br>
+     * </samp>
+     */
+    protected abstract void setValue(String key, String value);
+
+    /**
+     returns T.VALID_QUERY_PARAMETER_NVA_KEYS
+     */
+    protected Set<String> validKeys() {
+        return VALID_LUCENE_PARAMETER_KEYS;
+    }
+
+    protected boolean invalidQueryParameter(T key, String value) {
+        return isNull(value) || !value.matches(key.pattern());
+    }
+
+
+    protected Set<String> getMissingKeys() {
+        return
+            requiredMissing()
+                .stream()
+                .map(IParameterKey::key)
+                .collect(Collectors.toSet());
+    }
+
+    protected Set<T> required() {
+        return query.otherRequiredKeys;
+
+    }
+
+    protected Set<T> requiredMissing() {
+        return
+            required().stream()
+                .filter(key -> !query.luceneParameters.containsKey(key))
+                .filter(key -> !query.queryParameters.containsKey(key))
+                .collect(Collectors.toSet());
+    }
+
+    protected void throwInvalidParameterValue(Map.Entry<T, String> entry) throws BadRequestException {
```

Comment:
this says throw exception, but it does not say it also checks for validity before throwing exception. Should they be separate methods, or a better name?